src/components/tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,306 @@
+import uiUtilities from '@teamleader/ui-utilities';
+import cx from 'classnames';
+import omit from 'lodash.omit';
+import React, { MouseEventHandler, ReactNode, useEffect, useMemo, useRef, useState } from 'react';
+import { createPortal } from 'react-dom';
+import Transition from 'react-transition-group/Transition';
+import { GenericComponent } from '../../@types/types';
+import { COLORS, SIZES } from '../../constants';
+import Box from '../box';
+import { BoxProps } from '../box/Box';
+import DocumentObjectProvider, { Context as DocumentObjectContext } from '../hoc/DocumentObjectProvider';
+import { getViewport } from '../utils/utils';
+import theme from './theme.css';
+
+type Position = 'bottom' | 'horizontal' | 'left' | 'right' | 'top' | 'vertical';
+
+export const POSITIONS: Record<string, Position> = {
+  BOTTOM: 'bottom',
+  HORIZONTAL: 'horizontal',
+  LEFT: 'left',
+  RIGHT: 'right',
+  TOP: 'top',
+  VERTICAL: 'vertical',
+};
+
+interface PositionState {
+  position: Position;
+  top: number | string;
+  left: number | string;
+}
+export type AllowedColor = Exclude<typeof COLORS[number], 'teal'> | 'white' | 'inverse';
+export type AllowedSize = Exclude<typeof SIZES[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;
+const SIZE_MAP: Record<AllowedSize, BoxProps> = {
+  large: {
+    padding: 4,
+  },
+  medium: {
+    padding: 3,
+  },
+  small: {
+    paddingHorizontal: 3,
+    paddingVertical: 2,
+  },
+};
+
+interface TooltippedComponentProps {
+  children: ReactNode;
+  className?: string;
+  onClick?: MouseEventHandler;
+  onMouseEnter?: MouseEventHandler;
+  onMouseLeave?: MouseEventHandler;
+  onTooltipEntered?: () => void;
+  tooltip: ReactNode;
+  tooltipColor?: AllowedColor;
+  tooltipHideOnClick?: boolean;
+  tooltipIcon?: ReactNode;
+  tooltipPosition?: Position;
+  tooltipShowOnClick?: boolean;
+  tooltipSize?: AllowedSize;
+  documentObject: Document;
+  tooltipShowDelay?: number;
+  /** The z-index of the Tooltip */
+  zIndex?: number;
+  tooltipActive?: boolean;
+  ComposedComponent: React.ElementType;
+  // Because its HOC, it needs to pass props of child component
+  [key: string]: any;
```

Comment:
After many attempts, it seems like if you remove this and change the type of the `Tooltip` component you can have proper/better type checking/autocomplete

(Had to create signature for react elements and html elements)
```
function Tooltip<E extends keyof JSX.IntrinsicElements>(
  ComposedComponent: E,
): React.ComponentType<JSX.IntrinsicElements[E] & TooltipProps>;
function Tooltip<P>(ComposedComponent: React.ElementType<P>): React.ComponentType<P & TooltipProps>;

function Tooltip(ComposedComponent: React.ElementType) {
  return DocumentObjectProvider((props) => {
    ...
```

Reply:
Awesome 🔥